drafts/2019-03-15-integration-testing-with-testcontainers.adoc

```diff
@@ -0,0 +1,317 @@
+    @Path("/{personId}")
+    public Person getPerson(@PathParam("personId") long id) { /* ... */ }
+
+    @POST
+    public Long createPerson(@QueryParam("name") @NotEmpty @Size(min = 2, max = 50) String name,
+                             @QueryParam("age") @PositiveOrZero int age) { /* ... */ }
+
+    @PATCH
+    @Path("/{personId}")
+    public void updatePerson(@PathParam("personId") long id, @Valid Person p) { /* ... */ }
+
+    @DELETE
+    @Path("/{personId}")
+    public void removePerson(@PathParam("personId") long id) { /* ... */ }
+}
+----
+
+=== Leveraging application classes to simplify test client invocations
+
+If we have the application classes under test on the test client's classpath, we can do 2 interesting things:
+
+1. Build a type-safe rest client for invoking operations on `PersonService`
+2. Leverage JSON-B to automatically convert between `Person` objects and JSON data
+
+This is made possible by the following test-only dependencies:
+[source,groovy]
+----
+dependencies {
+  // JSON-B
+  testCompile group: 'org.eclipse', name: 'yasson', version: '1.0.3'
+  testCompile group: 'org.glassfish', name: 'javax.json', version: '1.1.4'
+  // Type-safe rest client
+  testCompile group: 'org.apache.cxf', name: 'cxf-rt-rs-client', version: '3.3.0'
+  testCompile group: 'cglib', name: 'cglib-nodep', version: '3.2.10'
+}
+----
+
+With these test dependencies, we can create and use a type-safe rest client like this:
+[source,java]
+----
+@ClassRule
+public static GenericContainer libertyContainer = new GenericContainer("my-service")
+                    .withExposedPorts(9080)
+                    .withLogConsumer(new Slf4jLogConsumer(LOGGER))
+                    .waitingFor(Wait.forHttp("/myservice"));
+
+private static PersonService personSvc; 
+
+@BeforeClass
+public static void setupClass() {
+    List<Class<?>> providers = new ArrayList<>();
+    providers.add(MyJsonBProvider.class); // trivial implementation of a JAX-RS MessageBodyReader/Writer
+    String appPath = "http://" + libertyContainer.getContainerIpAddress() + ':' + 
+                                 libertyContainer.getFirstMappedPort() + 
+                                 "/myservice"
+    personSvc = JAXRSClientFactory.create(appPath, PersonService.class, providers);
+}
+
+@Test
+public void testGetPerson() {
+    // invokes HTTP POST on http://<my-service-ip>:<my-service-port>/myservice/
+    // with query params name="Bob" age=24 (as defined on PersonService.createPerson())
+    Long bobId = personSvc.createPerson("Bob", 24);
+    // returns a generated ID
+
+    // invokes HTTP GET http://<my-service-ip>:<my-service-port>/myservice/{bobID}
+    // JSON response is automatically deserlialized into Person object
+    Person bob = personSvc.getPerson(bobId);
+
+    // Now that we have a POJO representing the response, verify data with simple JUnit assertions
+    assertEquals("Bob", bob.name);
+    assertEquals(24, bob.age);
+    assertNotNull(bob.id);
+}
+----
```

Comment:
That example is pretty slick!  Integrates very nicely with JUnit concepts that developers already know.